Move to Buildkite output

rikhuijzer opened the pull request:
This PR switches to Buildkite outputs.

Also, ensures that pull requests are **not** deployed.

rikhuijzer commented:
Okay. Great. This just deployed from a PR.

devmotion commented:
IIRC the tutorials, and hence this PR, are not ready yet. Eg due to some bugs in old CSV and Parsers versions some tutorials don't run successfully.

devmotion commented:
Maybe best to stick with the artifacts for one or two additional weeks.

rikhuijzer commented:
> IIRC the tutorials, and hence this PR, are not ready yet. Eg due to some bugs in old CSV and Parsers versions some tutorials don't run successfully.

Well, for what it's worth. The Turing 0.20 is now used for tutorial 1 :rofl: at https://turing.ml/dev/tutorials/00-introduction/ so your latest PR is published.

You can unsubscribe from this PR for now. I'll continue with disabling publishing for PRs because that makes no sense at all. Once that is fixed, I can force push the `gh-pages` branch back.

devmotion commented:
> The Turing 0.20 is now used for tutorial 1 🤣 at https://turing.ml/dev/tutorials/00-introduction/ so your latest PR is published.

Nice. And it even seems to work, also maths and figures. I was almost certain that I broke something in the transition 😂

rikhuijzer commented:
> Maybe best to stick with the artifacts for one or two additional weeks.

@devmotion by the way, you saw that the Documentation build is currently broken? See the GitHub Actions logs. Someone removed the artifacts branch from TuringTutorials, so waiting will block any deployments from this repo unless someone puts the artifacts branch back. I have a version locally but I don't know how new it is.

rikhuijzer commented:
> > The Turing 0.20 is now used for tutorial 1 rofl at [https://turing.ml/dev/tutorials/00-introduction/ so](https://turing.ml/dev/tutorials/00-introduction/%C2%A0so) your latest PR is published.
> 
> Nice. And it even seems to work, also maths and figures. I was almost certain that I broke something in the transition joy

Let's try to move forward with publishing then? I can help out with debugging from there

rikhuijzer commented:
The PR functionality is now fixed. It will not deploy anymore. From the logs:

```
┌ Info: Deployment criteria for deploying preview build from GitHub Actions:
│ - ✔ ENV["GITHUB_REPOSITORY"]="TuringLang/turing.ml" occurs in repo="github.com:TuringLang/turing.ml.git"
│ - ✔ ENV["GITHUB_REF"] corresponds to a PR number
│ - ✔ PR originates from the same repository
│ - ✘ `push_preview` keyword argument to deploydocs is `true`
│ - ✔ ENV["GITHUB_ACTOR"] exists and is non-empty
│ - ✔ ENV["DOCUMENTER_KEY"] exists and is non-empty
└ Deploying: ✘
```

for both the `Build and deploy (master)` and `Build and deploy (stable)` step.

I've also reverted the deploy in the `gh-pages` branch by force pushing it back.

I'll now leave this PR alone. It can be merged once the Buildkite outputs are satisfactory.

devmotion commented:
> Someone removed the artifacts branch from TuringTutorials, so waiting will block any deployments from this repo unless someone puts the artifacts branch back. I have a version locally but I don't know how new it is.

That's really bad. I didn't remove it, at least not on purpose. Sometimes the auto-delete bot messes up stuff but I don't think this happened here since I don't think we ever made a PR from the artifacts branch to another branch.

I checked, and currently tutorial 4 is missing (I think someone merged some PRs shortly after the transition PR was merged which canceled the buildkite job that was supposed to build and publish all tutorials) and tutorials 8, 10, 11, and 12 are broken, mainly/only due to outdated packages and code. I think it's bad to publish so many broken tutorials on the webpage, so until they are fixed it would good to keep the current state and possibly recover the artifacts branch. I think we should take the time and update and revise the tutorials properly and not stress us unnecessarily. Two weeks more or less don't matter, I guess.

devmotion commented on make-utils.jl:
```diff
@@ -1,6 +1,6 @@
 using Base64, Pkg
-Pkg.add(url="https://github.com/TuringLang/TuringTutorials", rev="artifacts")
-using TuringTutorials
+const OUTPUT_DIR = "TuringTutorialsOutput"
+run(`git clone --depth=1 https://github.com/TuringLang/TuringTutorialsOutput $OUTPUT_DIR`)
```
Probably better to use https://docs.julialang.org/en/v1/stdlib/LibGit2/#LibGit2.clone so we don't have to rely on an installed `git`:
```suggestion
using LibGit2
const OUTPUT_DIR = "TuringTutorialsOutput"
LibGit2.clone("https://github.com/TuringLang/TuringTutorialsOutput", $OUTPUT_DIR)
```

rikhuijzer replied:
Okay. I've switched to LibGit2. I'm always a bit hesitant with that because it doesn't allow clones with `--depth=1` (https://github.com/libgit2/libgit2/issues/3058). In this case, the difference was 7.2 seconds versus 7.9 seconds, so it isn't too bad.

devmotion commented:
I think we should be able to switch to TuringTutorialsOutput. All tutorials are built successfully. Of course, it's not perfect yet (suboptimal code, old design, deprecations, old packages, ...) but it is already an improvement compared with the previous, currently published version.

devmotion commented on make-utils.jl:
```diff
@@ -310,6 +312,8 @@ function copy_tutorial(tutorial_dest_path)
         # Clean up temporary workspace
         rm(tmp_path, recursive=true)
+
+        rm(OUTPUT_DIR, recursive=true)
```
Isn't
```suggestion
        rm(md_path, recursive=true)
```
better, in case one wants to rerun the function and/or avoid re-cloning the repo?

rikhuijzer replied:
Probably, but it's very hard to keep the repo in the right updated state so that's why I probably did it like this. As long as you don't restart Julia, it won't run again.

devmotion replied:
The Github action restarts Julia (runs the script twice for different versions of Turing), so I think it's preferable to not reclone the repo. Maybe it's good though to do a clone or update above, i.e., to update to the latest commit if the TuringTutorialsOutput repo is already checked out? Or is this already implied by `clone`? It does not matter for the action here but I guess it would be more convenient if you run the script locally.

rikhuijzer replied:
Could you finish the PR yourself? That is a much more efficient way than this back and forth messaging with hours in between.

rikhuijzer replied:
Or let's just merge and fix that nitpick later. The only thing in the way of a good PR is the dream of a perfect PR.

devmotion replied:
Well, actually I don't dare to finish, approve, and/or merge this PR. The main reason is that I'm not familiar with all workarounds, why they were added, and why they are required (or why they could be removed). Even more so since I almost missed that removing the deploy config and changing the action would have broken builds on the master branch. So I think this PR needs @cpfiffer's approval before merging it.

While looking at the action once more, I just noticed that in fact Julia is restarted and the repo will be recloned - which your comment above seemed to suggest would be the only case when removing it would be undesirable.

cpfiffer replied:
To be honest, I'm not going to be of much use here -- I'm in dissertation mode and I can't dedicate the time required to dig back into this stuff and figure out all the tape and glue that makes this thing work. 

If @rikhuijzer is fine with taking point here, I say we let him push it through and make changes as needed.

yebai replied:
I second @cpfiffer's suggestion here. It is not a big deal breaking some stuff here (everything is tracked and can be fixed with additional PR), particularly the docs building is already broken. 

@rikhuijzer would you be happy to push this through?

rikhuijzer commented on make.jl:
```diff
@@ -101,20 +101,13 @@ cp(joinpath(tmp_path, "assets"), joinpath(tmp_path, "_site", "assets"), force=tr
 
 repo = "github.com:TuringLang/turing.ml.git"
 
-deploy_config = GitHubActions(
-    "TuringLang/turing.ml", #github_repository::String
-    "push", #github_event_name::String
-    is_dev ? "refs/heads/master" : "refs/tags/$(ARGS[1])" #github_ref::String
-)
```
Removing this avoids publishing on PRs.

devmotion replied:
I don't think this can be removed - it seems this would break the current Github action setup that publishes both the stable and dev version of the docs in every run. A manual `deploy_config` is needed to be able to do this.

Generally, it would also be good to add previews for PRs to be able to review them. However, it this involves too many changes maybe it should be done in a separate PR.

devmotion reviewed (commented):
I think the changes of the deploy script are problematic (https://github.com/TuringLang/turing.ml/pull/26#discussion_r846848830). I'm not really familiar with the current deployment structure and assume that it could/should be improved (I guess this might be fixed in the new webpage). 

Maybe it would be better though to not change the action (Documentation.yml) or make.jl in this PR. Also running the action in PRs is of limited use if there is no preview of the webpage.

devmotion commented on make-utils.jl:
```diff
@@ -1,6 +1,9 @@
+using LibGit2
+using Pkg
+
+const OUTPUT_DIR = "TuringTutorialsOutput"
+clone("https://github.com/TuringLang/TuringTutorialsOutput", OUTPUT_DIR)
```
Coming back to the discussion in https://github.com/TuringLang/turing.ml/pull/26/files#r846601425 about removing `OUTPUT_DIR`: It seems `OUTPUT_DIR` is only used in this specific function below but not anywhere else. So it seems we could just remove the global constant and move `clone(...)` to that function? That would also solve any problems with rerunning the function (since currently the cloned repo is removed as discussed in the other thread).

devmotion replied:
@rikhuijzer If you rather want to just merge the PR, I can open a follow-up PR that removes the global constant.

devmotion commented:
:tada:
